Allow speculative parser to see inside Declarative Shadow DOM (DSD) templates

Ahmad-S792 opened the pull request:
#### 83e87b6e6a3beda3ed62ced43e7dc01aaf44d327
<pre>
Allow speculative parser to see inside Declarative Shadow DOM (DSD) templates

<a href="https://bugs.webkit.org/show_bug.cgi?id=269697">https://bugs.webkit.org/show_bug.cgi?id=269697</a>
<a href="https://rdar.apple.com/problem/123623233">rdar://problem/123623233</a>

Reviewed by Antti Koivisto.

Merge: <a href="https://chromium-review.googlesource.com/c/chromium/src/+/4200836">https://chromium-review.googlesource.com/c/chromium/src/+/4200836</a>

Prior to this patch, the speculative parser would bail out at the sight
of any `template` element. However, if that template was a
Declarative Shadow DOM shadow root, like `template shadowrootmode=open`,
it should be scanned.

Note that DSD shadow roots inside &quot;normal&quot; template elements
should still be skipped, since everything in the `template`
is inert.

* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::scan):
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-img-src.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-link-stylesheet.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-script-src.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-img-src.tentative-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-link-stylesheet.tentative-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-script-src.tentative-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/279668@main">https://commits.webkit.org/279668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/535de1d30538fcc541d44e48d76d7ff074af7b28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33496 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57387 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4730 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43823 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3220 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46847 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/24965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28549 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4163 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2985 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58981 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4471 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51238 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30486 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46961 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50601 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11792 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31449 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30266 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -467,8 +469,18 @@ void TokenPreloadScanner::scan(const HTMLToken& token, Vector<std::unique_ptr<Pr
+            }
+            // If this is a declarative shadow root <template shadowrootmode>
+            // element *and* we're not already inside a non-DSD <template> element,
+            // then we leave the template count at zero. Otherwise, increment it.
```
I don't know what acronym "DSD" means here or in the bug title. Please expand it.

Ahmad-S792 replied:
Sure, I will update bug title plus comment and then will do `merge-queue` directly.

Ahmad-S792 replied:
Done. @anttijk - Changed in comment, commit message and title.

Commit:
Allow speculative parser to see inside Declarative Shadow DOM (DSD) templates

https://bugs.webkit.org/show_bug.cgi?id=269697
rdar://problem/123623233

Reviewed by Antti Koivisto.

Merge: https://chromium-review.googlesource.com/c/chromium/src/+/4200836

Prior to this patch, the speculative parser would bail out at the sight
of any `template` element. However, if that template was a
Declarative Shadow DOM shadow root, like `template shadowrootmode=open`,
it should be scanned.

Note that DSD shadow roots inside "normal" template elements
should still be skipped, since everything in the `template`
is inert.

* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::scan):
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-img-src.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-link-stylesheet.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/template-shadowrootmode-script-src.tentative.sub-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-img-src.tentative-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-link-stylesheet.tentative-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/template-shadowrootmode-script-src.tentative-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/279668@main
